{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# General file editing\n",
    "\n",
    "## Purpose \n",
    "The following commands are used to edit files that aren't for one particular program. Associated scripts are found in the [General](https://github.com/brittanyhowell/sv_discovery/blob/master/general) folder of [SV_discovery](https://github.com/brittanyhowell/sv_discovery).\n",
    "\n",
    "## April 12\n",
    "\n",
    "### Make a bam index for all 4760 chr20 only bams\n",
    "Commit: [moved file](https://github.com/brittanyhowell/sv_discovery/blob/master/general/indexBAM.sh)\n",
    "```\n",
    "bsub -J \"BDArray[1-4760]\" -o /nfs/team151/bh10/scripts/breakdancer_bh10/output/indexBAM-%I-%J.out -e /nfs/team151/bh10/scripts/breakdancer_bh10/output/indexBAM-%I-%J.err \"/nfs/team151/bh10/scripts/breakdancer_bh10/indexBAM.sh /nfs/team151/bh10/scripts/genomestrip_bh10/fileLists/bamList.txt\"\n",
    "```\n",
    "\n",
    "\n",
    "## April 18\n",
    "\n",
    "### Filter bams based on standard chromosomes, run on WG bams\n",
    "Commit: [added reporting echoes](https://github.com/brittanyhowell/sv_discovery/blob/master/general/extract_std_chr.sh)\n",
    "```\n",
    "bsub -n3 -J  \"filter[1-226]\" -o /nfs/team151/bh10/scripts/bh10_general/output/filter/filter_std_chr%I-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/filter/filter_std_chr%I-%J.err -R\"select[mem>2000] rusage[mem=2000] span[hosts=1]\" -M2000 /nfs/team151/bh10/scripts/bh10_general/extract_std_chr.sh\n",
    "\n",
    "Job <9600455> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "## April 20\n",
    "\n",
    "### Index 225 WG Bams\n",
    "Commit: [rename file, change DIRs](https://github.com/brittanyhowell/sv_discovery/blob/master/general/indexSAM.sh)\n",
    "\n",
    "```\n",
    "bsub -J \"BDArray[1-225]\" -o /nfs/team151/bh10/scripts/bh10_general/output/indexBAM-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/indexBAM-%J-%I.err \"/nfs/team151/bh10/scripts/bh10_general/indexBAM.sh /lustre/scratch115/projects/interval_wgs/WGbams_std/WG_225_std.list\"\n",
    "\n",
    "Job <9667951> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "## April 26 \n",
    "\n",
    "### Convert 3724 WG CRAMs to BAMs\n",
    "Commit: [Changed for 3724 WG CRAMs](https://github.com/brittanyhowell/sv_discovery/blob/master/general/CRAMtoBAM.sh)\n",
    "\n",
    "```\n",
    "bsub -J \"cramArray[1-3724]\" -o /nfs/team151/bh10/scripts/bh10_general/output/CRAMtoBAM/CRAM-to-BAM-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/CRAMtoBAM/CRAM-to-BAM-%J-%I.err /nfs/team151/bh10/scripts/bh10_general/CRAMtoBAM.sh\n",
    "\n",
    "Job <218379> is submitted to default queue <normal>.\n",
    "```\n",
    "#### Failed: wrong list of files\n",
    "\n",
    "### Use correct file list!\n",
    "Commit: [correct file list this time I promise](https://github.com/brittanyhowell/sv_discovery/blob/master/general/CRAMtoBAM.sh)\n",
    "\n",
    "```\n",
    "bsub -J \"cramArray[1-3724]\" -o /nfs/team151/bh10/scripts/bh10_general/output/CRAMtoBAM/CRAM-to-BAM-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/CRAMtoBAM/CRAM-to-BAM-%J-%I.err /nfs/team151/bh10/scripts/bh10_general/CRAMtoBAM.sh\n",
    "\n",
    "Job <218702> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "#### Output: (.err)\n",
    "```\n",
    "cat *218702*.err \n",
    "```\n",
    "\n",
    "### Index WG BAMs \n",
    "\n",
    "Commit: [change for 3724 WG bams](https://github.com/brittanyhowell/sv_discovery/blob/master/general/indexBAM.sh)\n",
    "```\n",
    "bsub -J \"BDArray[1-3724]\" -o /nfs/team151/bh10/scripts/bh10_general/output/indexBAM-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/indexBAM-%J-%I.err \"/nfs/team151/bh10/scripts/bh10_general/indexBAM.sh /nfs/team151/bh10/scripts/bh10_general/fileLists/WG_bams.list\"\n",
    "\n",
    "Job <256158> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "#### Output: .out file\n",
    "```\n",
    "/nfs/team151/bh10/scripts/bh10_general/output - \n",
    "\n",
    "ls *.out | wc -l:  3724\n",
    "cat *.out | grep -c \"Successfully completed.\": 3724\n",
    "\n",
    "```\n",
    "Ah wait. Just because it \"sucessfully completed\" the job, doesn't mean the job successfully did what it was meant to do. \n",
    "\n",
    "#### Output: Any .err file\n",
    "```\n",
    "line 34: /lustre/scratch115/projects/interval_wgs/WGbams_std/EGAN00001345051.bam.bai: No such file or directory\n",
    "```\n",
    "Issue: I forgot to update the code on the FARM before it ran. So it still operated on the previous DIRs\n",
    "\n",
    "\n",
    "## April 27\n",
    "\n",
    "### Index WG BAMs\n",
    "Nothing changed, I forgot to update the code on the FARM. So it was looking for old DIRs. \n",
    "\n",
    "```\n",
    "bsub -J \"BDArray[1-3724]\" -o /nfs/team151/bh10/scripts/bh10_general/output/indexBAM-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/indexBAM-%J-%I.err \"/nfs/team151/bh10/scripts/bh10_general/indexBAM.sh /nfs/team151/bh10/scripts/bh10_general/fileLists/WG_bams.list\"\n",
    "\n",
    "Job <257181> is submitted to default queue <normal>.\n",
    "```\n",
    "#### Output: \n",
    "```\n",
    "cat indexBAM-257181-*.err - No printing.\n",
    "cat *.out | grep -c \"Successfully completed.\" : 3724. Nice. \n",
    "```\n",
    "\n",
    "\n",
    "## May 16\n",
    "\n",
    "### Intersect with GENCODE protein coding genes\n",
    "Considering I have a list of x000 samples, I think I am going to cat a long list of protein coding genes and then use it and then delete it. \n",
    "\n",
    "``` zcat chr*.gz | awk  '$5==\"protein_coding\"' > all_chr_protein_coding.txt ```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### run Intersect\n",
    "\n",
    "\n",
    "### Intersect\n",
    "\n",
    "#### intersect-genes.go\n",
    "Commit: [changed output file format to make merging more logical](https://github.com/brittanyhowell/sv_discovery/blob/master/general/intersect_genes.go)\n",
    "\n",
    "#### run-intersect-genes.sh\n",
    "Commit: [set up for array](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-intersect-genes.sh)\n",
    "```\n",
    "bsub -J \"int[1-3643]\" -o /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.err \"/nfs/team151/bh10/scripts/bh10_general/run-intersect-genes.sh\" \n",
    "\n",
    "Job <593957> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "### intersect again\n",
    "I made the output more verbose so I could investigate a few options and make sure it was working\n",
    "\n",
    "#### intersect-genes.go\n",
    "Commit: [more verbose output](https://github.com/brittanyhowell/sv_discovery/blob/master/general/intersect_genes.go)\n",
    "\n",
    "#### run-intersect-genes.sh\n",
    "Commit: [change outDIR](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-intersect-genes.sh)\n",
    "```\n",
    "bsub -J \"int[1-3643]\" -o /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.err \"/nfs/team151/bh10/scripts/bh10_general/run-intersect-genes.sh\" \n",
    "\n",
    "Job <594192> is submitted to default queue <normal>.\n",
    "```\n",
    "### Intersect, once more\n",
    "Honestly, there were many more attempts in the middle of these ones, but this one works.\n",
    "\n",
    "#### intersect-genes.go\n",
    "Commit: [The issue with sex chromosomes](https://github.com/brittanyhowell/sv_discovery/blob/master/general/intersect_genes.go)\n",
    "\n",
    "#### run-intersect-genes.sh\n",
    "Commit: [The issue with sex chromosomes](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-intersect-genes.sh)\n",
    "\n",
    "```\n",
    "bsub -n3  -J \"int[1-3642]\" -o /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.err  -R\"select[mem>2000] rusage[mem=2000] span[hosts=1]\" -M2000 \"/nfs/team151/bh10/scripts/bh10_general/run-intersect-genes.sh\" \n",
    "\n",
    "Job <598780> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "### Intersect again!\n",
    "\n",
    "No need to merge chr 23 with 24, it's both a waste of time *AND* innaccurate\n",
    " \n",
    " #### intersect-genes.go\n",
    "Commit: [The issue with sex chromosomes](https://github.com/brittanyhowell/sv_discovery/blob/master/general/intersect_genes.go)\n",
    "\n",
    "#### run-intersect-genes_BD.sh\n",
    "Commit: [remove chr merge](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-intersect-genes_BD.sh)\n",
    "\n",
    "\n",
    "```\n",
    "bsub -n3  -J \"int[1-3642]\" -o /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/intersectBD/intersect-%J-%I.err  -R\"select[mem>2000] rusage[mem=2000] span[hosts=1]\" -M2000 \"/nfs/team151/bh10/scripts/bh10_general/run-intersect-genes.sh\" \n",
    "Job <600494> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "## May 17 \n",
    "\n",
    "### Merge files \n",
    "\n",
    "merges all intersects into one super power table to rule all tables.\n",
    "\n",
    "I regret my commit. Why. \n",
    "#### run_merge_intersectGenes.sh\n",
    "Commit: [hi i run a thing](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run_merge_intersectGenes.sh)\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/merge-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/merge-%I.err -R\"select[mem>1000] rusage[mem=1000]\" -M1000 /nfs/team151/bh10/scripts/bh10_general/mergeTab.sh\n",
    "\n",
    "Job <711848> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "## May 18\n",
    "\n",
    "Resubmit with more memory: \n",
    "\n",
    "#### run_merge_intersectGenes.sh\n",
    "Commit: [hi i run a thing](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run_merge_intersectGenes.sh)\n",
    "\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.err -R\"select[mem>4000] rusage[mem=4000]\" -M4000 /nfs/team151/bh10/scripts/bh10_general/mergeTab.sh\n",
    "\n",
    "Job <725988> is submitted to default queue <normal>.\n",
    "```\n",
    "Failed after merging 400. I have over 3000. I may not have enough memory. \n",
    "\n",
    "### Hey lets try this one.\n",
    "\n",
    "I'll look up a new way to merge the tables. For now, throw more memory at it. \n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.err -R\"select[mem>20000] rusage[mem=20000]\" -M20000 /nfs/team151/bh10/scripts/bh10_general/mergeTab.sh\n",
    "\n",
    "Job <726617> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "Failed after merging 2018 - it ran out of time.\n",
    "\n",
    "\n",
    "## May 21\n",
    "### Run intersect.go on non genotyped discovery data\n",
    "\n",
    "#### \n",
    "run-intersect-genes_GS_no_geno.sh\n",
    "Commit: [we run this](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-intersect-genes_GS_no_geno.sh)\n",
    "\n",
    "#### intersect_genes.go\n",
    "Commit: [The issue with sex chromosomes](https://github.com/brittanyhowell/sv_discovery/blob/master/general/intersect_genes.go)\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/intersectGS/intersect-%J-%I.out -e bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/intersectGS/intersect-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/intersectGS/intersect-%J-%I.err  -R\"select[mem>1000] rusage[mem=1000]\" -M1000  \"/nfs/team151/bh10/scripts/bh10_general/run-intersect-genes_GS.sh\" \n",
    "\n",
    "Job <911058> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "### Run intersect.go on genotyped CNV data\n",
    "\n",
    "#### run-intersect-genes_GS_geno.sh\n",
    "Commit: [we run this](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-intersect-genes_GS_geno.sh)\n",
    "\n",
    "#### intersect_genes.go\n",
    "Commit: [The issue with sex chromosomes](https://github.com/brittanyhowell/sv_discovery/blob/master/general/intersect_genes.go)\n",
    "\n",
    "```\n",
    "bsub -J \"int[1-226]\"  -o /nfs/team151/bh10/scripts/bh10_general/output/intersectGS/geno/intersect-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/intersectGS/geno/intersect-%J-%I.err  -R\"select[mem>1000] rusage[mem=1000]\" -M1000  \"/nfs/team151/bh10/scripts/bh10_general/run-intersect-genes_GS.sh\" \n",
    "\n",
    "Job <923014> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "```\n",
    "grep \"Successfully completed.\" *.out | wc -l : 226\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "## May 22\n",
    "\n",
    "### run-filter_repeats.sh\n",
    "\n",
    "run the repeat filter over the discovery set \n",
    "\n",
    "####  run-filter_repeats.sh\n",
    "Commit: [arguments added - suited to run on disc. GS set](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run-filter_repeats.sh)\n",
    "\n",
    "#### filter-repeats.R\n",
    "Commit: [arguments added - suited to run on disc. GS set](https://github.com/brittanyhowell/sv_discovery/blob/master/general/filter-repeats.R)\n",
    "\n",
    "```\n",
    " bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/repeatFilter/repFilt-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/repeatFilter/repFilt-%J-%I.err  -R\"select[mem>1000] rusage[mem=1000]\" -M1000  \"/nfs/team151/bh10/scripts/bh10_general/run-filter_repeats.sh\" \n",
    "\n",
    "Job <933669> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "### Run again, with more memory\n",
    "\n",
    "Upgraded from 1GB to 3GB\n",
    "```\n",
    " bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/repeatFilter/repFilt-%J-%I.out -e /nfs/team151/bh10/scripts/bh10_general/output/repeatFilter/repFilt-%J-%I.err  -R\"select[mem>3000] rusage[mem=3000]\" -M3000  \"/nfs/team151/bh10/scripts/bh10_general/run-filter_repeats.sh\" \n",
    "\n",
    "Job <935170> is submitted to default queue <normal>.\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "## May 22\n",
    "\n",
    "### Merge small - GS CNV\n",
    "Run merge on th 226 GS samples - this should work, and is useful for the association study. If only for now.\n",
    "\n",
    "#### run_merge_intersectGenes.sh\n",
    "Commit:[run merge on 226 samples, rather than thousands. Might work](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run_merge_intersectGenes.sh)\n",
    "\n",
    "#### merge_intersectGenes.R\n",
    "Commit:[run merge on 226 samples, rather than thousands. Might work](https://github.com/brittanyhowell/sv_discovery/blob/master/general/merge_intersectGenes.R)\n",
    "\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.err -R\"select[mem>10000] rusage[mem=10000]\" -M10000 /nfs/team151/bh10/scripts/bh10_general/mergeTab.sh\n",
    "\n",
    "Job <949105> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "### Extract 226 BD samples that match the 226 GS ones\n",
    "\n",
    "#### extract_226.sh\n",
    "Commit:[extract the 226 GS files from the pool of BD samples](https://github.com/brittanyhowell/sv_discovery/blob/master/general/extract_226.sh)\n",
    "\n",
    "\n",
    "### Merge small - BD\n",
    "Run merge on th 226 BD samples - this should work, the last one did.\n",
    "\n",
    "#### run_merge_intersectGenes.sh\n",
    "Commit:[run merge on 226 samples, rather than thousands. Might work](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run_merge_intersectGenes.sh)\n",
    "\n",
    "#### merge_intersectGenes.R\n",
    "Commit:[changed for BD merge](https://github.com/brittanyhowell/sv_discovery/blob/master/general/merge_intersectGenes.R)\n",
    "\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.err -R\"select[mem>10000] rusage[mem=10000]\" -M10000 /nfs/team151/bh10/scripts/bh10_general/mergeTab.sh\n",
    "\n",
    "Job <951389> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "Commit: [why. Why did I forget the D](https://github.com/brittanyhowell/sv_discovery/blob/master/general/merge_intersectGenes.R)\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/merge-%J.err -R\"select[mem>10000] rusage[mem=10000]\" -M10000 /nfs/team151/bh10/scripts/bh10_general/mergeTab.sh\n",
    "\n",
    "Job <953274> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "\n",
    "### Association - CNV\n",
    "\n",
    "#### extract_for_association.R\n",
    "Commit: [CNV association run 1](https://github.com/brittanyhowell/sv_discovery/blob/master/general/extract_for_association.R)\n",
    "\n",
    "#### association.R\n",
    "Commit: [CNV association run 1](https://github.com/brittanyhowell/sv_discovery/blob/master/general/association.R)\n",
    "\n",
    "#### run_CNV_association.sh\n",
    "Commit: [CNV association run 1](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run_CNV_association.sh)\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/associate_CNV-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/associate_CNV-%J.err -R\"select[mem>1000] rusage[mem=1000]\" -M1000 /nfs/team151/bh10/scripts/bh10_general/run-association_CNV.sh\n",
    "\n",
    "Job <973832> is submitted to default queue <normal>.\n",
    "\n",
    "```\n",
    "\n",
    "### Forgot about the trailing args..\n",
    "\n",
    "\n",
    "#### extract_for_association.R\n",
    "Commit: [added comments and trailing args arguments](https://github.com/brittanyhowell/sv_discovery/blob/master/general/extract_for_association.R)\n",
    "\n",
    "#### association.R\n",
    "Commit: [added comments and trailing args arguments](https://github.com/brittanyhowell/sv_discovery/blob/master/general/association.R)\n",
    "\n",
    "#### run_CNV_association.sh\n",
    "Commit: [added comments and trailing args arguments](https://github.com/brittanyhowell/sv_discovery/blob/master/general/run_CNV_association.sh)\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    " bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/associate_CNV-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/associate_CNV-%J.err -R\"select[mem>1000] rusage[mem=1000]\" -M1000 /nfs/team151/bh10/scripts/bh10_general/run-association_CNV.sh\n",
    "\n",
    "Job <974347> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "#### run with more memory - it ran out\n",
    "\n",
    "```\n",
    "bsub  -o /nfs/team151/bh10/scripts/bh10_general/output/associate_CNV-%J.out -e /nfs/team151/bh10/scripts/bh10_general/output/associate_CNV-%J.err -R\"select[mem>4000] rusage[mem=4000]\" -M4000 /nfs/team151/bh10/scripts/bh10_general/run-association_CNV.sh\n",
    "\n",
    "Job <1000339> is submitted to default queue <normal>.\n",
    "```\n",
    "\n",
    "## May 23\n",
    "### Extract 226 samples of BD for the intersect\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
